src/apply.ts

```diff
@@ -26,154 +24,93 @@ interface ReadOptions {
 	application: number;
 	authToken: string;
 	stream: stream.Readable;
-	force?: boolean;
 }
 
-function generateUniqueKey() {
-	return uuidv4().replace(/-/g, '');
+interface Release {
+	releaseImages: Array<{
+		image: SDK.Image;
+		service: string;
+	}>;
 }
 
-async function backfillLowerRevisions(
-	sdk: SDK.BalenaSDK,
-	application: number,
-	currentDateIso: string,
-	manifest: ReleaseManifest,
-): Promise<void> {
-	const [highestSameSemverLocalReleases] = await sdk.pine.get({
-		resource: 'release',
-		options: {
-			$top: 1,
-			$select: ['id', 'revision'],
-			$filter: {
-				belongs_to__application: application,
-				semver_major: manifest.semver_major,
-				semver_minor: manifest.semver_minor,
-				semver_patch: manifest.semver_patch,
-				revision: { $ne: null },
-			},
-			$orderby: 'revision desc',
-		},
-	});
+function normalizeManifest(manifest: SDK.Release): Release {
+	const release: Release = { releaseImages: [] };
 
-	const version = `${manifest.semver_major}.${manifest.semver_minor}.${manifest.semver_patch}`;
-	for (
-		//
-		let revision =
-			highestSameSemverLocalReleases == null
-				? 0
-				: // we have already $filtered for this so ! is fine
-					highestSameSemverLocalReleases.revision! + 1;
-		revision < (manifest.revision ? manifest.revision : 0);
-		revision++
-	) {
-		await sdk.pine.post({
-			resource: 'release',
-			body: {
-				belongs_to__application: application,
-				status: 'failed',
-				start_timestamp: currentDateIso,
-				end_timestamp: currentDateIso,
-				update_timestamp: currentDateIso,
-				source: manifest.source,
-				composition: manifest.composition,
-				commit: generateUniqueKey(),
-				contract: manifest.contract,
-				semver: revision === 0 ? version : `${version}+rev${revision}`,
-			},
+	if (!Array.isArray(manifest.release_image)) {
+		throw new Error(
+			`Expected array of release images but found ${typeof manifest.release_image}`,
+		);
+	}
+	for (const releaseImage of manifest.release_image) {
+		const image = releaseImage.image;
+		if (!Array.isArray(image)) {
+			throw new Error(
+				`Expected array of images in release image ${releaseImage.id} but found ${typeof image}`,
+			);
+		}
+		const service = image[0].is_a_build_of__service;
+		if (!Array.isArray(service)) {
+			throw new Error(
+				`Expected array of services in release image ${releaseImage.id} but found ${typeof service}`,
+			);
+		}
+		if (image[0].status !== 'success' || image[0].content_hash == null) {
+			throw new Error(
+				`Expected a status of 'success' for release image ${releaseImage.id} but found status ${image[0].status}`,
```

Comment:
What if status is success but content hash null?

Reply:
Right, we can't just assume that both conditions will always be met since the release bundle can still be modified externally.  This should be separate value checks.